### PR TITLE
Refactor column prune rule and LogicalProject

### PR DESCRIPTION
**Pull request by HangyuanLiu**
#493

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 1]
```diff
@@ -65,57 +72,32 @@ public OptExpression optimize(ConnectContext connectContext,
...
-        context.getTaskScheduler().pushTask(new TopDownRewriteTask(rootTaskContext,
-                memo.getRootGroup(), RuleSetType.PARTITION_PRUNE));
-        context.getTaskScheduler().executeTasks(rootTaskContext, memo.getRootGroup());
-
-        context.getTaskScheduler().pushTask(new TopDownRewriteTask(rootTaskContext,
-                memo.getRootGroup(), RuleSetType.SCALAR_OPERATOR_REUSE));
-        context.getTaskScheduler().executeTasks(rootTaskContext, memo.getRootGroup());
+        ruleRewriteOnlyOnce(memo, rootTaskContext, RuleSetType.PARTITION_PRUNE);
+        ruleRewriteIterative(memo, rootTaskContext, new PruneProjectRule());
+        ruleRewriteIterative(memo, rootTaskContext, new ScalarOperatorsReuseRule());
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:100]: Expression reuse should run only once？

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 2]
```diff
@@ -65,57 +72,32 @@ public OptExpression optimize(ConnectContext connectContext,
...
-        rootTaskContext.setRequiredColumns((ColumnRefSet) requiredColumns.clone());
-        context.getTaskScheduler().pushTask(new TopDownRewriteTask(rootTaskContext,
-                memo.getRootGroup(), RuleSetType.PRUNE_COLUMNS));
-        context.getTaskScheduler().executeTasks(rootTaskContext, memo.getRootGroup());
-
-        context.getTaskScheduler().pushTask(new TopDownRewriteTask(rootTaskContext,
-                memo.getRootGroup(), RuleSetType.PRUNE_ASSERT_ROW));
-        context.getTaskScheduler().executeTasks(rootTaskContext, memo.getRootGroup());
+        ruleRewriteIterative(memo, rootTaskContext, new MergeTwoProjectRule());
+        ruleRewriteIterative(memo, rootTaskContext, RuleSetType.PRUNE_ASSERT_ROW);
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:92]: Maybe we can use a MEREG_OPERATOR group to describe these rules?

Comment (HangyuanLiu) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:92]: Merge and other rules are in order and cannot be executed in a group

**Comment by HangyuanLiu** [comment]
run starrocks_fe_unittest

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PredicatePrune.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+// select partitions:
+//  PARTITION p3 VALUES [2020-04-01, 2020-07-01)
+//  PARTITION p4 VALUES [2020-07-01, 2020-12-01)
+//
+// predicate:
+//  d = 2020-02-02 AND d > 2020-08-01, None prune
+//  d >= 2020-04-01 AND d > 2020-09-01, All Prune
+//  d >= 2020-04-01 AND d < 2020-09-01, "d >= 2020-04-01" prune, "d < 2020-09-01" not prune
+//  d IN (2020-05-01, 2020-06-01), None prune
+public class PredicatePrune extends TransformationRule {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PredicatePrune.java:48]:
```suggestion
public class PartitionPredicatePrune extends TransformationRule {
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 4]
```diff
@@ -219,4 +201,22 @@ private void collectAllScanOperators(Memo memo, TaskContext rootTaskContext) {
         Utils.extractOlapScanOperator(tree.getGroupExpression(), list);
         rootTaskContext.setAllScanOperators(Collections.unmodifiableList(list));
     }
+
+    void ruleRewriteIterative(Memo memo, TaskContext rootTaskContext, RuleSetType ruleSetType) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:205]: Add some comments let other people know when to call `ruleRewriteIterative`, when should call `ruleRewriteOnlyOnce`

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/TopDownIterativeTask.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.util.List;
+
+/**
+ * TopDownIterativeTask performs a top-down rewrite logic operator pass.
+ * <p>
+ * This class modify from CMU noisepage project TopDownRewrite class
+ */
+public class TopDownIterativeTask extends OptimizerTask {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/TopDownIterativeTask.java:22]: This class could merge to `TopDownRewriteTask`

**Review by Seaven — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PruneWindowRule.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.logical.LogicalWindowOperator;
+import com.starrocks.sql.optimizer.operator.pattern.Pattern;
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.util.Collections;
+import java.util.List;
+
+public class PruneWindowRule extends TransformationRule {
+    public PruneWindowRule() {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PruneWindowRule.java:15]: Why we need this new rule, why not us the old way in PruneWindowColumnsRule

**Comment by HangyuanLiu** [comment]
run starrocks_fe_unittest

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java (hunk 4 above)
Comment (HangyuanLiu) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:205]: OK

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/TopDownIterativeTask.java (hunk 5 above)
Comment (HangyuanLiu) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/TopDownIterativeTask.java]: done

**Review by Youngwb — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**